src/wire/ipv6hop.rs

```diff
@@ -0,0 +1,307 @@
+    ///
+    /// The result of this check is invalidated by calling [set_header_len].
+    ///
+    /// [set_header_len]: #method.set_header_len
+    pub fn check_len(&self) -> Result<()> {
+        let data = self.buffer.as_ref();
+        let len = data.len();
+
+        if len < field::MIN_HEADER_SIZE {
+            return Err(Error::Truncated);
+        }
+
+        let of = field::OPTIONS(data[field::LENGTH]);
+
+        if len < of.end {
+            return Err(Error::Truncated);
+        }
+
+        Ok(())
+    }
+
+    /// Consume the packet, returning the underlying buffer.
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Return the next header field.
+    #[inline]
+    pub fn next_header(&self) -> Protocol {
+        let data = self.buffer.as_ref();
+        Protocol::from(data[field::NXT_HDR])
+    }
+
+    /// Return the header length field. Length of the Hop-by-Hop Options header in 8-octet units,
+    /// not including the first 8 octets.
+    #[inline]
+    pub fn header_len(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        data[field::LENGTH]
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> Packet<&'a T> {
+    /// Return the option data.
+    #[inline]
+    pub fn data(&self) -> &'a[u8] {
+        let data = self.buffer.as_ref();
+        &data[field::OPTIONS(data[field::LENGTH])]
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> Packet<T> {
+    /// Set the next header field.
+    #[inline]
+    pub fn set_next_header(&mut self, value: Protocol) {
+        let data = self.buffer.as_mut();
+        data[field::NXT_HDR] = value.into();
+    }
+
+    /// Set the option data length. Length of the Hop-by-Hop Options header in 8-octet units,
+    /// not including the first 8 octets.
+    #[inline]
+    pub fn set_header_len(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        data[field::LENGTH] = value;
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + AsMut<[u8]> + ?Sized> Packet<&'a mut T> {
+    /// Return a mutable pointer to the option data.
+    #[inline]
+    pub fn data_mut(&mut self) -> &mut [u8] {
+        let data = self.buffer.as_mut();
+        let len = data[field::LENGTH];
+        &mut data[field::OPTIONS(len)]
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> fmt::Display for Packet<&'a T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match Repr::parse(self) {
+            Ok(repr) => write!(f, "{}", repr),
+            Err(err) => {
+                write!(f, "IPv6 Hop-by-Hop Options ({})", err)?;
+                Ok(())
+            }
+        }
+    }
+}
+
+/// A high-level representation of an IPv6 Hop-by-Hop Options header.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct Repr {
+    /// The type of header immediately following the Hop-by-Hop Options header.
+    pub next_header: Protocol,
+    /// Length of the Hop-by-Hop Options header in 8-octet units, not including the first 8 octets.
+    pub length:      u8,
```

Comment:
Shouldn't this mention option data instead?

Reply:
Not sure I follow. Are you saying that the length in `Repr` should be the length of the options data? Or are you saying the that the `Repr` should contain options data. If the latter, we discussed this a bit [here](https://github.com/m-labs/smoltcp/pull/139#discussion_r165240227).

Reply:
Yeah, I wasn't sure if this `Repr` would be more usable if it was like the ICMP, TCP, and UDP reprs where the `data` is in the `Repr`, or like like IPv6 and IPv4 where the `data` is not in the `Repr`s.

Reply:
@whitequark should the `Repr` contain `data` or should it be outside of it?